xSQLServerAlwaysOnService: isHadrEnabled null in Get-TargetResource

I'm running into an issue where xSQLServerAlwaysOnService fails even though it is correctly setting the isHadrEnabled setting to True.  After it sets the value and restarts, it calls Test-TargetResource again, which first calls Get-TargetResource.  In Get-TargetResource, the verbose message throws an "Index operation failed; the array index evaluated to null." when it tries to show the status of Hadr.  Interestingly, it falls back through to Test-TargetResource which then *is* able to see the same value as True.  I have no idea why this is happening, but it kills the DSC at this point and I can't continue with the AlwaysOn setup.  Any help would be appreciated.  My DSC log output is below.  This is version 6.0.0.0 of xSQLServer on Windows 2012 R2 and SQL Server 2014 SP2.

```
VERBOSE: [2017-04-19T22:07:22] [VERBOSE] [myserver]: LCM:  [ Start  Resource ]  
[[xSQLServerAlwaysOnService]AlwaysOnService::[AlwaysOnInstall]InstallSQL]
VERBOSE: [2017-04-19T22:07:22] [VERBOSE] [myserver]: LCM:  [ Start  Test     ]  
[[xSQLServerAlwaysOnService]AlwaysOnService::[AlwaysOnInstall]InstallSQL]
VERBOSE: [2017-04-19T22:07:22] [VERBOSE] [myserver]:                            
[[xSQLServerAlwaysOnService]AlwaysOnService::[AlwaysOnInstall]InstallSQL] 2017-04-19_22-07-22: Connected to SQL 
myserver\myinstance
VERBOSE: [2017-04-19T22:07:23] [VERBOSE] [myserver]:                            
[[xSQLServerAlwaysOnService]AlwaysOnService::[AlwaysOnInstall]InstallSQL] 2017-04-19_22-07-23: SQL Always On is disabled on 
"myserver\myinstance".
VERBOSE: [2017-04-19T22:07:23] [VERBOSE] [myserver]:                            
[[xSQLServerAlwaysOnService]AlwaysOnService::[AlwaysOnInstall]InstallSQL] 2017-04-19_22-07-23: SQL Always On is in the desired 
state for "myserver\myinstance": False.
VERBOSE: [2017-04-19T22:07:23] [VERBOSE] [myserver]: LCM:  [ End    Test     ]  
[[xSQLServerAlwaysOnService]AlwaysOnService::[AlwaysOnInstall]InstallSQL]  in 0.0620 seconds.
VERBOSE: [2017-04-19T22:07:23] [VERBOSE] [myserver]: LCM:  [ Start  Set      ]  
[[xSQLServerAlwaysOnService]AlwaysOnService::[AlwaysOnInstall]InstallSQL]
VERBOSE: [2017-04-19T22:07:23] [VERBOSE] [myserver]:                            
[[xSQLServerAlwaysOnService]AlwaysOnService::[AlwaysOnInstall]InstallSQL] 2017-04-19_22-07-23: Importing SQLPS module.
VERBOSE: [2017-04-19T22:07:27] [VERBOSE] [myserver]:                            
[[xSQLServerAlwaysOnService]AlwaysOnService::[AlwaysOnInstall]InstallSQL] 2017-04-19_22-07-27: Enabling Always On for the 
instance myserver\myinstance
VERBOSE: [2017-04-19T22:07:28] [VERBOSE] [myserver]:                            
[[xSQLServerAlwaysOnService]AlwaysOnService::[AlwaysOnInstall]InstallSQL] 2017-04-19_22-07-28: SQL Always On has been enabled 
on "myserver\myinstance". Restarting the service.
VERBOSE: [2017-04-19T22:07:28] [VERBOSE] [myserver]:                            
[[xSQLServerAlwaysOnService]AlwaysOnService::[AlwaysOnInstall]InstallSQL] 2017-04-19_22-07-28: Connected to SQL 
myserver\myinstance
VERBOSE: [2017-04-19T22:07:28] [VERBOSE] [myserver]:                            
[[xSQLServerAlwaysOnService]AlwaysOnService::[AlwaysOnInstall]InstallSQL] 2017-04-19_22-07-28: Getting SQL Service information
VERBOSE: [2017-04-19T22:07:28] [VERBOSE] [myserver]:                            
[[xSQLServerAlwaysOnService]AlwaysOnService::[AlwaysOnInstall]InstallSQL] 2017-04-19_22-07-28: SQL Server service restarting
VERBOSE: [2017-04-19T22:07:35] [VERBOSE] [myserver]:                            
[[xSQLServerAlwaysOnService]AlwaysOnService::[AlwaysOnInstall]InstallSQL] 2017-04-19_22-07-35: Starting SQL Server Agent (myinstance)
VERBOSE: [2017-04-19T22:07:35] [VERBOSE] [myserver]:                            
[[xSQLServerAlwaysOnService]AlwaysOnService::[AlwaysOnInstall]InstallSQL] 2017-04-19_22-07-35: Connected to SQL 
myserver\myinstance
VERBOSE: [2017-04-19T22:07:38] **[ERROR] Index operation failed; the array index evaluated to null.**
VERBOSE: [2017-04-19T22:07:39] [VERBOSE] [myserver]:                            
[[xSQLServerAlwaysOnService]AlwaysOnService::[AlwaysOnInstall]InstallSQL] 2017-04-19_22-07-39: SQL Always On is in the desired 
state for "myserver\myinstance": True.
VERBOSE: [2017-04-19T22:07:39] [VERBOSE] [myserver]: LCM:  [ End    Set      ]  
[[xSQLServerAlwaysOnService]AlwaysOnService::[AlwaysOnInstall]InstallSQL]  in 16.0790 seconds.
VERBOSE: [2017-04-19T22:07:39] [ERROR] The PowerShell DSC resource 
'[xSQLServerAlwaysOnService]AlwaysOnService::[AlwaysOnInstall]InstallSQL' with SourceInfo 'C:\Program 
Files\WindowsPowerShell\Modules\AlwaysOnCompositeDSC\DSCResources\AlwaysOnInstall\AlwaysOnInstall.schema.psm1::786::13::xSQLSer
verAlwaysOnService' threw one or more non-terminating errors while running the Set-TargetResource functionality. These errors 
are logged to the ETW channel called Microsoft-Windows-DSC/Operational. Refer to this channel for more details.
VERBOSE: [2017-04-19T22:07:39] [VERBOSE] [myserver]: LCM:  [ End    Set      ]
VERBOSE: [2017-04-19T22:07:39] [ERROR] The SendConfigurationApply function did not succeed.
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -23,7 +23,7 @@
 - Changes to xSQLServer
   - Fixed typos in markdown files; CHANGELOG, CONTRIBUTING, README and ISSUE_TEMPLATE.
 - Changes to xSQLServerAlwaysOnService
-  - Get-TargetResource should no longer fail silently with error 'Index operation failed; the array index evaluated to null.'. Now if the Server.IsHadrEnabled property return neither $true or $false the Get-TargetResource function will throw an error.
+  - Get-TargetResource should no longer fail silently with error 'Index operation failed; the array index evaluated to null.' (issue #519). Now if the Server.IsHadrEnabled property return neither $true or $false the Get-TargetResource function will throw an error.
 
 ## 7.0.0.0
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -22,6 +22,8 @@
   - Removed helper function Get-SQLPSInstance and Get-SQLPSInstanceName because there is no resource using it any longer.
 - Changes to xSQLServer
   - Fixed typos in markdown files; CHANGELOG, CONTRIBUTING, README and ISSUE_TEMPLATE.
+- Changes to xSQLServerAlwaysOnService
+  - Get-TargetResource should no longer fail silently with error 'Index operation failed; the array index evaluated to null.'. Now if the Server.IsHadrEnabled property return neither $true or $false the Get-TargetResource function will throw an error.
 
 ## 7.0.0.0
 

--- a/DSCResources/MSFT_xSQLServerAlwaysOnService/MSFT_xSQLServerAlwaysOnService.psm1
+++ b/DSCResources/MSFT_xSQLServerAlwaysOnService/MSFT_xSQLServerAlwaysOnService.psm1
@@ -9,12 +9,12 @@ Import-Module -Name (Join-Path -Path (Split-Path (Split-Path $PSScriptRoot -Pare
     .PARAMETER Ensure
     *** Not used in this function ***
     HADR is Present (enabled) or Absent (disabled).
-    
+
     .PARAMETER SQLServer
     Hostname of the SQL Server to be configured.
-    
+
     .PARAMETER SQLInstanceName
-    Name of the SQL instance to be configued. 
+    Name of the SQL instance to be configued.
 #>
 function Get-TargetResource
 {
@@ -36,11 +36,44 @@ function Get-TargetResource
         $SQLInstanceName
     )
 
-    $sql = Connect-SQL -SQLServer $SQLServer -SQLInstanceName $SQLInstanceName
+    $sqlServerObject = Connect-SQL -SQLServer $SQLServer -SQLInstanceName $SQLInstanceName
 
-    New-VerboseMessage -Message ( 'SQL Always On is {0} on "{1}\{2}".' -f @{$false='disabled'; $true='enabled'}[$sql.IsHadrEnabled],$SQLServer,$SQLInstanceName )
+    $isAlwaysOnEnabled = $sqlServerObject.IsHadrEnabled
+    if ($isAlwaysOnEnabled -eq $true)
+    {
+        $statusString = 'enabled'
+    }
+    elseif ($isAlwaysOnEnabled -eq $false)
+    {
+        $statusString = 'disabled'
+    }
+    else
+    {
+        # This is a validation test for issue #519.
+        try {
+            if ($isAlwaysOnEnabled -eq $null)
+            {
+               throw 'Server.IsHadrEnabled was set to $null.'
+            }
+            else
+            {
+                $statusString = $isAlwaysOnEnabled
 
-    return @{ IsHadrEnabled = $sql.IsHadrEnabled }
+                throw 'Server.IsHadrEnabled was set to unexpected value.'
+            }
+        }
+        catch
+        {
+            throw New-TerminatingError -ErrorType UnexpectedAlwaysOnStatus -FormatArgs $statusString -ErrorCategory InvalidResult -InnerException $_.Exception
+        }
+    }
+
+
+    New-VerboseMessage -Message ( 'SQL Always On is {0} on "{1}\{2}".' -f $statusString, $SQLServer, $SQLInstanceName )
+
+    return @{
+        IsHadrEnabled = $isAlwaysOnEnabled
+    }
 }
 
 <#
@@ -52,7 +85,7 @@ function Get-TargetResource
 
     .PARAMETER SQLServer
     Hostname of the SQL Server to be configured.
-    
+
     .PARAMETER SQLInstanceName
     Name of the SQL instance to be configued.
 
@@ -128,10 +161,10 @@ function Set-TargetResource
 
     .PARAMETER Ensure
     HADR is Present (enabled) or Absent (disabled).
-    
+
     .PARAMETER SQLServer
     Hostname of the SQL Server to be configured.
-    
+
     .PARAMETER SQLInstanceName
     Name of the SQL instance to be configued.
 
@@ -162,7 +195,7 @@ function Test-TargetResource
         [Int32]
         $RestartTimeout = 120
     )
-    
+
     # Determine the current state of Always On
     $params = @{
         Ensure = $Ensure
@@ -171,8 +204,8 @@ function Test-TargetResource
     }
 
     $state = Get-TargetResource @params
-    
-    # Determine what the desired state of Always On is 
+
+    # Determine what the desired state of Always On is
     $hadrDesiredState = @{ 'Present' = $true; 'Absent' = $false }[$Ensure]
 
     # Determine whether the value matches the desired state

--- a/Tests/Unit/MSFT_xSQLServerAlwaysOnService.Tests.ps1
+++ b/Tests/Unit/MSFT_xSQLServerAlwaysOnService.Tests.ps1
@@ -257,7 +257,7 @@ try
             }
         }
 
-        Context 'When HADR is not in the desired state' {
+        Context 'When HADR is in the desired state' {
             Mock -CommandName Connect-SQL -MockWith {
                 return New-Object PSObject -Property @{
                     IsHadrEnabled = $true

--- a/Tests/Unit/MSFT_xSQLServerAlwaysOnService.Tests.ps1
+++ b/Tests/Unit/MSFT_xSQLServerAlwaysOnService.Tests.ps1
@@ -1,7 +1,7 @@
 $script:DSCModuleName      = 'xSQLServer'
 $script:DSCResourceName    = 'MSFT_xSQLServerAlwaysOnService'
 
-# Unit Test Template Version: 1.1.0 
+# Unit Test Template Version: 1.1.0
 [String] $script:moduleRoot = Split-Path -Parent (Split-Path -Parent $PSScriptRoot)
 if ( (-not (Test-Path -Path (Join-Path -Path $script:moduleRoot -ChildPath 'DSCResource.Tests'))) -or `
      (-not (Test-Path -Path (Join-Path -Path $script:moduleRoot -ChildPath 'DSCResource.Tests\TestHelper.psm1'))) )
@@ -13,7 +13,7 @@ Import-Module (Join-Path -Path $script:moduleRoot -ChildPath 'DSCResource.Tests\
 $TestEnvironment = Initialize-TestEnvironment `
     -DSCModuleName $script:DSCModuleName `
     -DSCResourceName $script:DSCResourceName `
-    -TestType Unit 
+    -TestType Unit
 
 $disableHadr = @{
     Ensure = 'Absent'
@@ -47,26 +47,26 @@ try
         Context 'When HADR is disabled' {
 
             Mock -CommandName Connect-SQL -MockWith {
-                return New-Object PSObject -Property @{ 
+                return New-Object PSObject -Property @{
                     IsHadrEnabled = $false
                 }
             } -ModuleName $script:DSCResourceName -Verifiable
 
             It 'Should return that HADR is disabled' {
-                
+
                 # Get the current state
                 $result = Get-TargetResource @enableHadr
-                
+
                 $result.IsHadrEnabled | Should Be $false
 
                 Assert-MockCalled -ModuleName $script:DSCResourceName -CommandName Connect-SQL -Scope It -Times 1 -Exactly
             }
 
             It 'Should return that HADR is disabled' {
-                
+
                 # Get the current state
                 $result = Get-TargetResource @enableHadrNamedInstance
-                
+
                 $result.IsHadrEnabled | Should Be $false
 
                 Assert-MockCalled -ModuleName $script:DSCResourceName -CommandName Connect-SQL -Scope It -Times 1 -Exactly
@@ -76,28 +76,56 @@ try
         Context 'When HADR is enabled' {
 
             Mock -CommandName Connect-SQL -MockWith {
-                return New-Object PSObject -Property @{ 
+                return New-Object PSObject -Property @{
                     IsHadrEnabled = $true
                 }
             } -ModuleName $script:DSCResourceName -Verifiable
 
             It 'Should return that HADR is enabled' {
-                
+
                 # Get the current state
                 $result = Get-TargetResource @enableHadr
-                
+
                 $result.IsHadrEnabled | Should Be $true
 
                 Assert-MockCalled -ModuleName $script:DSCResourceName -CommandName Connect-SQL -Scope It -Times 1 -Exactly
             }
 
             It 'Should return that HADR is enabled' {
-                
+
                 # Get the current state
                 $result = Get-TargetResource @enableHadrNamedInstance
-                
+
                 $result.IsHadrEnabled | Should Be $true
 
+                Assert-MockCalled -ModuleName $script:DSCResourceName -CommandName Connect-SQL -Scope It -Times 1 -Exactly
+            }
+        }
+
+        # This it regression test for issue #519.
+        Context 'When Server.IsHadrEnabled returns $null' {
+            Mock -CommandName Connect-SQL -MockWith {
+                return New-Object PSObject -Property @{
+                    IsHadrEnabled = $null
+                }
+            } -ModuleName $script:DSCResourceName -Verifiable
+
+            It 'Should fail with the correct error message' {
+                { Get-TargetResource @enableHadr } | Should Not Throw 'Index operation failed; the array index evaluated to null'
+                { Get-TargetResource @enableHadr } | Should Throw 'The status of property Server.IsHadrEnabled was netiher $true or $false. Status is ''''. InnerException: Server.IsHadrEnabled was set to $null.'
+                Assert-MockCalled -ModuleName $script:DSCResourceName -CommandName Connect-SQL -Scope It -Times 2 -Exactly
+            }
+        }
+
+        Context 'When Server.IsHadrEnabled returns unexpected value' {
+            Mock -CommandName Connect-SQL -MockWith {
+                return New-Object PSObject -Property @{
+                    IsHadrEnabled = 'UnknownStatus'
+                }
+            } -ModuleName $script:DSCResourceName -Verifiable
+
+            It 'Should fail with the correct error message' {
+                { Get-TargetResource @enableHadr } | Should Throw 'The status of property Server.IsHadrEnabled was netiher $true or $false. Status is ''UnknownStatus''. InnerException: Server.IsHadrEnabled was set to unexpected value.'
                 Assert-MockCalled -ModuleName $script:DSCResourceName -CommandName Connect-SQL -Scope It -Times 1 -Exactly
             }
         }
@@ -113,23 +141,23 @@ try
         Mock -CommandName Enable-SqlAlwaysOn -MockWith {} -ModuleName $script:DSCResourceName
 
         Mock -CommandName Import-SQLPSModule -MockWith {} -ModuleName $script:DSCResourceName
-        
+
         Mock -CommandName New-TerminatingError { $ErrorType } -ModuleName $script:DSCResourceName
 
         Mock -CommandName New-VerboseMessage -MockWith {} -ModuleName $script:DSCResourceName
-        
+
         Mock -CommandName Restart-SqlService -MockWith {} -ModuleName $script:DSCResourceName -Verifiable
 
         Context 'When HADR is not in the desired state' {
 
             It 'Should enable SQL Always On when Ensure is set to Present' {
-                
+
                 Mock -CommandName Connect-SQL -MockWith {
-                    return New-Object PSObject -Property @{ 
+                    return New-Object PSObject -Property @{
                         IsHadrEnabled = $true
                     }
                 } -ModuleName $script:DSCResourceName -Verifiable
-                
+
                 Set-TargetResource @enableHadr
                 Assert-MockCalled -ModuleName $script:DSCResourceName -CommandName Connect-SQL -Scope It -Times 1
                 Assert-MockCalled -ModuleName $script:DSCResourceName -CommandName Disable-SqlAlwaysOn -Scope It -Times 0
@@ -138,13 +166,13 @@ try
             }
 
             It 'Should disable SQL Always On when Ensure is set to Absent' {
-                
+
                 Mock -CommandName Connect-SQL -MockWith {
-                return New-Object PSObject -Property @{ 
+                return New-Object PSObject -Property @{
                         IsHadrEnabled = $false
                     }
                 } -ModuleName $script:DSCResourceName -Verifiable
-                
+
                 Set-TargetResource @disableHadr
                 Assert-MockCalled -ModuleName $script:DSCResourceName -CommandName Connect-SQL -Scope It -Times 1
                 Assert-MockCalled -ModuleName $script:DSCResourceName -CommandName Disable-SqlAlwaysOn -Scope It -Times 1
@@ -153,13 +181,13 @@ try
             }
 
             It 'Should enable SQL Always On on a named instance when Ensure is set to Present' {
-                
+
                 Mock -CommandName Connect-SQL -MockWith {
-                    return New-Object PSObject -Property @{ 
+                    return New-Object PSObject -Property @{
                         IsHadrEnabled = $true
                     }
                 } -ModuleName $script:DSCResourceName -Verifiable
-                
+
                 Set-TargetResource @enableHadrNamedInstance
                 Assert-MockCalled -ModuleName $script:DSCResourceName -CommandName Connect-SQL -Scope It -Times 1
                 Assert-MockCalled -ModuleName $script:DSCResourceName -CommandName Disable-SqlAlwaysOn -Scope It -Times 0
@@ -168,13 +196,13 @@ try
             }
 
             It 'Should disable SQL Always On on a named instance when Ensure is set to Absent' {
-                
+
                 Mock -CommandName Connect-SQL -MockWith {
-                    return New-Object PSObject -Property @{ 
+                    return New-Object PSObject -Property @{
                         IsHadrEnabled = $false
                     }
                 } -ModuleName $script:DSCResourceName -Verifiable
-                
+
                 Set-TargetResource @disableHadrNamedInstance
                 Assert-MockCalled -ModuleName $script:DSCResourceName -CommandName Connect-SQL -Scope It -Times 1
                 Assert-MockCalled -ModuleName $script:DSCResourceName -CommandName Disable-SqlAlwaysOn -Scope It -Times 1
@@ -183,13 +211,13 @@ try
             }
 
             It 'Should throw the correct error message when Ensure is set to Present, but IsHadrEnabled is $false' {
-                
+
                 Mock -CommandName Connect-SQL -MockWith {
-                    return New-Object PSObject -Property @{ 
+                    return New-Object PSObject -Property @{
                         IsHadrEnabled = $false
                     }
                 } -ModuleName $script:DSCResourceName -Verifiable
-                
+
                 { Set-TargetResource @enableHadr } | Should Throw 'AlterAlwaysOnServiceFailed'
                 Assert-MockCalled -ModuleName $script:DSCResourceName -CommandName Connect-SQL -Scope It -Times 1
                 Assert-MockCalled -ModuleName $script:DSCResourceName -CommandName Disable-SqlAlwaysOn -Scope It -Times 0
@@ -199,13 +227,13 @@ try
             }
 
             It 'Should throw the correct error message when Ensure is set to Absent, but IsHadrEnabled is $true' {
-                
+
                 Mock -CommandName Connect-SQL -MockWith {
-                    return New-Object PSObject -Property @{ 
+                    return New-Object PSObject -Property @{
                         IsHadrEnabled = $true
                     }
                 } -ModuleName $script:DSCResourceName -Verifiable
-                
+
                 { Set-TargetResource @disableHadr } | Should Throw 'AlterAlwaysOnServiceFailed'
                 Assert-MockCalled -ModuleName $script:DSCResourceName -CommandName Connect-SQL -Scope It -Times 1
                 Assert-MockCalled -ModuleName $script:DSCResourceName -CommandName Disable-SqlAlwaysOn -Scope It -Times 1
@@ -217,19 +245,28 @@ try
     }
 
     Describe "$($script:DSCResourceName)\Test-TargetResource" {
-        
-        Mock -CommandName Connect-SQL -MockWith {
-            return New-Object PSObject -Property @{ 
-                IsHadrEnabled = $true
+        Context 'When HADR is not in the desired state' {
+            Mock -CommandName Connect-SQL -MockWith {
+                return New-Object PSObject -Property @{
+                    IsHadrEnabled = $true
+                }
+            } -ModuleName $script:DSCResourceName -Verifiable
+
+            It 'Should cause Test-TargetResource to return false when not in the desired state' {
+                Test-TargetResource @disableHadr | Should be $false
             }
-        } -ModuleName $script:DSCResourceName -Verifiable
-        
-        It 'Should cause Test-TargetResource to return false when not in the desired state' {
-            Test-TargetResource @disableHadr | Should be $false
         }
 
-        It 'Should cause Test-TargetResource to return true when in the desired state' {
-            Test-TargetResource @enableHadr | Should be $true
+        Context 'When HADR is not in the desired state' {
+            Mock -CommandName Connect-SQL -MockWith {
+                return New-Object PSObject -Property @{
+                    IsHadrEnabled = $true
+                }
+            } -ModuleName $script:DSCResourceName -Verifiable
+
+            It 'Should cause Test-TargetResource to return true when in the desired state' {
+                Test-TargetResource @enableHadr | Should be $true
+            }
         }
     }
 }

--- a/en-US/xSQLServer.strings.psd1
+++ b/en-US/xSQLServer.strings.psd1
@@ -48,6 +48,7 @@ ConfigurationRestartRequired = Configuration option '{0}' has been updated, but 
 
 # AlwaysOnService
 AlterAlwaysOnServiceFailed = Failed to ensure Always On is {0} on the instance '{1}'.
+UnexpectedAlwaysOnStatus = The status of property Server.IsHadrEnabled was netiher $true or $false. Status is '{0}'.
 
 # Login
 PasswordValidationFailed = Creation of the login '{0}' failed due to the following error: {1}

